Mathlib/CategoryTheory/MorphismProperty/Comma.lean

```diff
@@ -0,0 +1,243 @@
+
+@[simp]
+lemma comp_hom {X Y Z : P.Comma L R Q W} (f : X ⟶ Y) (g : Y ⟶ Z) :
+    (f ≫ g).hom = f.hom ≫ g.hom := rfl
+
+/-- If `i` is an isomorphism in `Comma L R`, it is also a morphism in `P.Comma L R Q W`. -/
+@[simps hom]
+def homFromCommaOfIsIso [Q.RespectsIso] [W.RespectsIso] {X Y : P.Comma L R Q W}
+    (i : X.toComma ⟶ Y.toComma) [IsIso i] :
+    X ⟶ Y where
+  __ := i
+  prop_hom_left := Q.of_isIso i.left
+  prop_hom_right := W.of_isIso i.right
+
+instance [Q.RespectsIso] [W.RespectsIso] {X Y : P.Comma L R Q W} (i : X.toComma ⟶ Y.toComma)
+    [IsIso i] : IsIso (homFromCommaOfIsIso i) := by
+  constructor
+  use homFromCommaOfIsIso (inv i)
+  constructor <;> ext : 1 <;> simp
+
+/-- Any isomorphism between objects of `P.Comma L R Q W` in `Comma L R` is also an isomorphism
+in `P.Comma L R Q W`.  -/
+@[simps]
+def isoFromComma [Q.RespectsIso] [W.RespectsIso] {X Y : P.Comma L R Q W}
+    (i : X.toComma ≅ Y.toComma) : X ≅ Y where
+  hom := homFromCommaOfIsIso i.hom
+  inv := homFromCommaOfIsIso i.inv
+
+/-- Constructor for isomorphisms in `P.Comma L R Q W` from isomorphisms of the left and right
+components and naturality in the forward direction. -/
+@[simps!]
+def isoMk [Q.RespectsIso] [W.RespectsIso] {X Y : P.Comma L R Q W} (l : X.left ≅ Y.left)
+    (r : X.right ≅ Y.right) (h : L.map l.hom ≫ Y.hom = X.hom ≫ R.map r.hom := by aesop_cat) :
+    X ≅ Y :=
+  isoFromComma (CategoryTheory.Comma.isoMk l r h)
+
+variable (L R P Q W)
+
+/-- The forgetful functor. -/
+@[simps]
+def forget : P.Comma L R Q W ⥤ Comma L R where
+  obj X := X.toComma
+  map f := f.hom
+
+instance : (forget L R P Q W).Faithful where
+  map_injective := Comma.Hom.ext'
+
+variable {L R P Q W}
+
+instance {X Y : P.Comma L R Q W} (f : X ⟶ Y) [IsIso f] : IsIso f.hom :=
+  (forget L R P Q W).map_isIso f
+
+lemma hom_homFromCommaOfIsIso [Q.RespectsIso] [W.RespectsIso] {X Y : P.Comma L R Q W}
+    (i : X ⟶ Y) [IsIso i.hom] :
+    homFromCommaOfIsIso i.hom = i :=
+  rfl
+
+lemma inv_hom {X Y : P.Comma L R Q W} (f : X ⟶ Y) [IsIso f] : (inv f).hom = inv f.hom := by
+  apply IsIso.eq_inv_of_hom_inv_id
+  rw [← comp_hom, IsIso.hom_inv_id, id_hom]
+
+variable (L R P Q W)
+
+instance [Q.RespectsIso] [W.RespectsIso] : (forget L R P Q W).ReflectsIsomorphisms where
+  reflects f hf := by
+    simp only [forget_obj, forget_map] at hf
+    rw [← hom_homFromCommaOfIsIso f]
+    infer_instance
+
+/-- The forgetful functor from the full subcategory of `Comma L R` defined by `P` is
+fully faithful. -/
+def forgetFullyFaithful : (forget L R P ⊤ ⊤).FullyFaithful where
+  preimage {X Y} f := ⟨f, trivial, trivial⟩
+
+instance : (forget L R P ⊤ ⊤).Full :=
+  Functor.FullyFaithful.full (forgetFullyFaithful L R P)
+
+end Comma
+
+end Comma
+
+section Over
+
+variable {T : Type*} [Category T]
+
+/-- Given a morphism property `P` on a category `C` and an object `X : C`, this is the
+subcategory of `Over X` defined by `P` where morphisms satisfy `Q`. -/
```

Comment:
Is this the right way around?
Or is this "subcategory of `Over X` defined by `Q` where morphisms satisfy `P`"

Reply:
This is the subcategory of `Over X` where the structural morphisms satisfy `P` and for `X Y : P.Over Q X` a morphism from `X` to `Y` is a morphism `f` in `Over X` where `f.left` satisfies `Q`. Does this clarify it?

Reply:
Oh then the original wording is right. I was just confused.